llvm/lib/Target/AMDGPU/AMDGPUOclcReflect.cpp

```diff
@@ -0,0 +1,91 @@
+//===- AMDGPUOclcReflect.cpp ----------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This pass searches for occurences of the AMDGPU_OCLC_REFLECT function, and
+// replaces the calls with some val dependent on the operand of the func. This
+// can be used to reflect across different implementations of functions at
+// compile time based on a compiler flag or some other means. The first use case
+// is to choose a safe or unsafe version of atomic_xor at compile time, which
+// can be chosen at compile time by setting the flag
+// --amdgpu-oclc-unsafe-int-atomics=true.
+//
+// This pass is similar to the NVPTX pass NVVMReflect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "AMDGPU.h"
+#include "llvm/IR/Constants.h"
+#include "llvm/IR/Dominators.h"
+#include "llvm/IR/InstIterator.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/Pass.h"
+#include "llvm/Support/CommandLine.h"
+
+using namespace llvm;
+
+#define AMDGPU_OCLC_REFLECT "__oclc_amdgpu_reflect"
+
+static cl::opt<bool>
+    AMDGPUReflectEnabled("amdgpu-oclc-reflect-enable", cl::init(true),
```

Comment:
Do we expect anyone to use this flag?

Reply:
A implementer could potentially use this flag so that amdgpu reflect is deferred until some later linking, as we do with libclc. Do you think I should add extra documentation for usage somewhere?

Reply:
Makes sense, thanks. I don't think we necessarily need to document it anywhere - it seems fairly obvious (lots of such options dotted around LLVM passes)